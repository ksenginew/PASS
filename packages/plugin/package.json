{
  "name": "@passlang/plugin",
  "version": "0.0.1",
  "description": "Plugin of Pass CSS preprocessor for Vite, Rollup, Webpack, and more...",
  "source": "src/index.js",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./vite": {
      "require": "./dist/vite.js",
      "import": "./dist/vite.mjs"
    },
    "./webpack": {
      "require": "./dist/webpack.js",
      "import": "./dist/webpack.mjs"
    },
    "./rollup": {
      "require": "./dist/rollup.js",
      "import": "./dist/rollup.mjs"
    },
    "./nuxt": {
      "require": "./dist/nuxt.js",
      "import": "./dist/nuxt.mjs"
    },
    "./types": {
      "require": "./dist/types.js",
      "import": "./dist/types.mjs"
    },
    "./*": "./*"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "rollup -c -w"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksenginew/pass.git"
  },
  "keywords": [
    "style",
    "core",
    "pass",
    "passlang",
    "preprocessor",
    "css",
    "unplugin",
    "vite",
    "webpack",
    "rollup"
  ],
  "author": "Kavindu Santhusa",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/ksenginew/pass/issues"
  },
  "homepage": "https://github.com/ksenginew/pass#readme",
  "dependencies": {
    "@rollup/pluginutils": "^5.0.2",
    "passlang": "workspace:^0.0.1",
    "unplugin": "^1.4.0"
  },
  "devDependencies": {
    "rollup-plugin-styles": "^4.0.0",
    "@passlang/core": "workspace:^0.0.1"
  }
}