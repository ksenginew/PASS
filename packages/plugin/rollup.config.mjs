import resolve from "@rollup/plugin-node-resolve";
import commonjs from "@rollup/plugin-commonjs";
import terser from "@rollup/plugin-terser";
import pkg from "./package.json" assert {type: "json"};

const banner = `/*!
 * ${pkg.name}
 * ${pkg.description}
 *
 * @version v${pkg.version}
 * @author ${pkg.author}
 * @homepage ${pkg.homepage}
 * @repository ${pkg.repository?.url}
 * @license ${pkg.license}
 */`;

const external = [
  ...Object.keys(pkg.optionalDependencies || {}),
  ...Object.keys(pkg.peerDependencies || {}),
  ...Object.keys(pkg.dependencies || {}),
];

export default {
  external: external,
  input: pkg.source,
  output: [
    pkg.main && { banner, file: pkg.main, format: "cjs", exports: "auto" },
    pkg.module && { banner, file: pkg.module, format: "es", exports: "auto" },
    pkg["umd:main"] && {
      banner,
      file: pkg["umd:main"],
      format: "umd",
      name: pkg.name.replace(/-./g, (match) => match[1].toUpperCase()),
      exports: "auto",
      globals: Object.fromEntries(
        external.map((dep) => [
          dep,
          dep.slice(1).replace(/-./g, (m) => m[1].toUpperCase()),
        ])
      ),
      plugins: [terser()],
    },
  ],
  plugins: [resolve(), commonjs()],
};
