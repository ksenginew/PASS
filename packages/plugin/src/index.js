import { createUnplugin } from "unplugin";
import { createFilter } from "@rollup/pluginutils";
import { compileString } from "passlang";

export default createUnplugin(
  /**
   * @param {{
   *   include?: import("@rollup/pluginutils").FilterPattern,
   *   exclude?: import("@rollup/pluginutils").FilterPattern,
   *   emit?: string,
   * }} options
   */
  ({ include = /\.pass\.[mc]?[jt]sx?$/, exclude, emit } = {}) => {
    let filter = createFilter(include, exclude);
    return {
      name: "pass",
      transformInclude(id) {
        return filter(id);
      },
      async transform(src, id) {
        let code = await compileString(src, id);
        if (emit) {
          this.emitFile({
            fileName: emit,
            name: emit,
            source: code,
            type: "asset",
          });
          return "";
        } else return code;
      },
    };
  },
);
