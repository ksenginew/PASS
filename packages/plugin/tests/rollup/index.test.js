import { rollup } from "rollup";
import plugin from "../../src/index.js";

describe("rollup plugin", () => {
  test("should work", async () => {
    let bundle = await rollup({
      input: "packages/plugin/tests/rollup/main.js",
      plugins: [plugin.rollup({ emit: "bundle.css" })],
    });
    let { output } = await bundle.generate({ format: "es" });
    expect(
      output
        .map(
          (file) =>
            `${file.fileName}\n\n${
              file.type === "chunk" ? file.code : file.source
            }`,
        )
        .join("\n-------------------------------------------------\n"),
    ).toMatchSnapshot();
  });
});
