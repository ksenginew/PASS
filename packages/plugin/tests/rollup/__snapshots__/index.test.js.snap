// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup plugin should work 1`] = `
"main.js

var main = 'nothing';

export { main as default };



bundle.css


  .main-header {
    background-color: #a2b9bc;
  }

  .menu-left {
    background-color: #b2ad7f;
  }

  .menu-right {
    background-color: #878f99;
  }
"
`;
